### PR TITLE
[Issue #5003] Description notification

### PR DESCRIPTION
**Pull request by babebe**
Fixes https://github.com/HHS/simpler-grants-gov/issues/5003

Update content generation for Description section per [DOC](https://docs.google.com/document/d/1ypejg8hL7uOXdrI7auwf9SWQ1Ag07o7rCjnrVCLfA-s/edit?tab=t.0#heading=h.sl8i9r9yex9j)
Add tests

This PR introduces `beautifulsoup4` to handle HTML truncation instead of implementing a custom truncation function.
Using BeautifulSoup avoids reinventing complex parsing logic and reduces long-term maintenance risk.
It allows us to: 
Truncate based on text content while preserving valid HTML structure
Avoid edge cases around nested elements, inline tags, and mixed content

Attempt at creating custom truncation utility, required:
- Proper HTML parsing
- Handling nested tags correctly
- Ensuring valid output after truncation
- Maintaining and testing edge cases over time

The additional dependency with this change is small and widely adopted.

### CODE REVIEW

File: api/src/task/notifications/opportunity_notifcation.py [hunk 1]
```diff
@@ -274,6 +274,16 @@ def _get_last_notified_versions(
 
         return {(row.user_id, row.opportunity_id): row[2] for row in results}
 
+    def _build_description_fields_content(self, description_change: dict, opp_id: int) -> str:
+        after = description_change["after"]
+        if after:
+            description_section = SECTION_STYLING.format("Description")
+            if len(after) > TRUNCATION_THRESHOLD:
+                after += f"<a href='{self.notification_config.frontend_base_url}/opportunity/{opp_id}' style='color:blue;'>...Read full description</a>"
+            description_section += f"{BULLET_POINTS_STYLING} <i>New Description:</i> {after}<br>"
```
Comment (chouinar) [inline api/src/task/notifications/opportunity_notifcation.py]: A technical problem that I don't know an answer to (maybe ask frontend in our engineer channel) - how do we truncate HTML?

The description field has HTML in it most of the time, if we truncate it - we'll be putting half-written HTML into our HTML, potentially partially through a tag. For example, on this random opportunity: https://simpler.grants.gov/opportunity/359874 it uses bold. If we happened to truncate in the middle of that text, suddenly the rest of our email is all bold.

Maybe we want to escape the HTML or remove the tags? I don't know - but we'll definitely run into some problems if we don't do anything.

Comment (babebe) [reply api/src/task/notifications/opportunity_notifcation.py]: @doug-s-nava . Any suggestions ?

Comment (doug-s-nava) [reply api/src/task/notifications/opportunity_notifcation.py]: so... we had to deal with this on the frontend as well. I built a function to handle truncating and being smart about closing tags. Reimplementing this in python would likely be gnarly, but maybe that's our best option?

https://github.com/HHS/simpler-grants-gov/blob/53d4ec408d501b984e33aa618a326ef4fc963d6b/frontend/src/utils/generalUtils.ts#L12

Comment (mdragon) [reply api/src/task/notifications/opportunity_notifcation.py]: For version one and initial launch, can we just not truncate the content? Or not show the content changes just note that "The description was changed." On our current timeline we don't really have time to implement HTML striping or closing tags automatically before we want to be sending emails?

@eric-serrato95 Ok just saying "The Description was changed" without showing what changed like we do for other fields for now?

Comment (eric-serrato95) [reply api/src/task/notifications/opportunity_notifcation.py]: @mdragon I think that's ok for launch, we can figure out how to handle large description changes for example after launch

File: api/src/task/notifications/opportunity_notifcation.py [hunk 2]
```diff
@@ -274,6 +289,71 @@ def _get_last_notified_versions(
 
         return {(row.user_id, row.opportunity_id): row[2] for row in results}
 
+    def _contains_regex(self, value: str, regex: str) -> bool:
+        return bool(re.search(regex, value))
+
+    def _truncate_html_safe(self, html_str: str) -> str:
```
Comment (chouinar) [inline api/src/task/notifications/opportunity_notifcation.py]: Could you move these to the string_utils file? They generic enough for reuse.

File: api/tests/src/task/notifications/test_opportunity_notification.py [hunk 3]
```diff
@@ -790,6 +801,94 @@ def test_build_eligibility_content(
...
         assert res == expected_html
 
+    @pytest.mark.parametrize(
+        "html_str,expected_html",
+        [
+            # Truncate mid-text inside inline tag <strong>, no closing tags
+            (
+                "<p>This is a <strong>very big description here!!!!",
+                "<p>This is a <strong>very big description here!!!!</strong>",
+            ),
```
Comment (chouinar) [inline api/tests/src/task/notifications/test_opportunity_notification.py]: Could you add a test for multiple of the same tag nested? Something like `<div>first<div>second layer<div>yet another layer</div></div></div>`

File: api/src/task/notifications/opportunity_notifcation.py [hunk 4]
```diff
@@ -274,6 +289,71 @@ def _get_last_notified_versions(
...
+                else:
+                    # Add non self-closing tags
+                    tag_name = re.match(r"<(\w+)", token)
+                    if tag_name and not token.endswith("/>"):
+                        tag_stack.append(tag_name.group(1))
+
+        # Remove block-level closing tags from the end
+        while output_tokens:
+            last = output_tokens[-1]
+            closing_match = re.match(r"</(\w+)>", last)
```
Comment (chouinar) [inline api/src/task/notifications/opportunity_notifcation.py]: For any of these regexes, especially one we pass to compile, could we define them statically with a note of what they're looking for? You don't necessarily need to go piece-by-piece like https://github.com/HHS/simpler-grants-gov/blob/main/api/src/logging/pii.py#L60 but at least something like "words ending in a slash" or whatever.

File: api/tests/src/task/notifications/test_opportunity_notification.py (hunk 3 above)
Comment (babebe) [reply api/tests/src/task/notifications/test_opportunity_notification.py]: Added

File: api/src/task/notifications/opportunity_notifcation.py [hunk 5]
```diff
@@ -276,11 +278,28 @@ def _get_last_notified_versions(
...
+            SECTION_STYLING.format("Description"),
+            f"{BULLET_POINTS_STYLING} <i>New Description:</i>",
+        ]
+
+        if len(after) > TRUNCATION_THRESHOLD:
+            truncated = after[:TRUNCATION_THRESHOLD]
+
+            if contains_regex(truncated, r"<[^>]+>"):
+                truncated = truncate_html_safe(truncated)
+
```
Comment (chouinar) [inline api/src/task/notifications/opportunity_notifcation.py]: Might be a naming thing, but should `truncate_html_safe` take in the truncation length?

I'm imagining I'd want to use it like so:
```py
my_html = "..."

truncated_html = truncate_html_safe(my_html, TRUNCATION_THRESHOLD)
```
And move the bit you have here to that utility function.

File: api/tests/src/task/notifications/test_opportunity_notification.py [hunk 6]
```diff
@@ -787,13 +794,85 @@ def test_build_eligibility_content(
...
+        self,
+        db_session,
+        html_str,
+        expected_html,
+        set_env_var_for_email_notification_config,
+        set_truncation_threshold,
+    ):
+
+        res = truncate_html_safe(html_str)
+        assert res == expected_html
```
Comment (chouinar) [inline api/tests/src/task/notifications/test_opportunity_notification.py]: This test should be in test_string_utils

**Review by kmk142789 — APPROVED** [review]

File: api/src/util/string_utils.py [hunk 7]
```diff
@@ -1,5 +1,20 @@
...
+    "div",
+    "section",
+    "article",
+    "header",
+    "footer",
+    "main",
+    "nav",
+    "aside",
+    "blockquote",
+    "ul",
```
Comment (joshtonava) [inline api/src/util/string_utils.py]: **comment:** one additional tag I can think of is `li`, there may be more. I want to confirm this list is exhaustive.

File: api/tests/src/task/notifications/test_opportunity_notification.py
Comment (joshtonava) [inline api/tests/src/task/notifications/test_opportunity_notification.py]: issue: Test is failing I believe here 
```
TypeError: OpportunityNotificationTask._build_description_fields_content() missing 1 required positional argument: 'opp_id'
```

File: api/src/util/string_utils.py (hunk 7 above)
Comment (joshtonava) [reply api/src/util/string_utils.py]: I think missing block tags are

li, ol, h1, h2, h3, h4, h5, h6, pre, hr, table, thead, tbody, tr, td, th, form, fieldset

File: api/src/util/string_utils.py [hunk 8]
```diff
@@ -20,3 +35,88 @@ def is_valid_uuid(value: str) -> bool:
...
+        i -= 1
+
+    return html
+
+
+def _truncate_preserving_html(html_str: str) -> str:
+    tag_stack: list = []  # Keep track of open tags
+    output_tokens: list = []  # list of tokens
+
+    # Match any tags or text
```
Comment (joshtonava) [inline api/src/util/string_utils.py]:
```suggestion
    # Match any HTML tag or text content between tags
    # Pattern breakdown:
    # - <[^>]+>  : Matches complete HTML tags (opening, closing, or self-closing)
    # - |        : OR operator
    # - [^<]+    : Matches any text content (non-tag characters)
```

File: api/src/util/string_utils.py [hunk 9]
```diff
@@ -20,3 +35,88 @@ def is_valid_uuid(value: str) -> bool:
...
+            return html
+        if html[i] == "<":
+            # Found start of a tag with no closing '>' → strip it
+            return html[:i]
+        i -= 1
+
+    return html
+
+
+def _truncate_preserving_html(html_str: str) -> str:
```
Comment (joshtonava) [inline api/src/util/string_utils.py]: **suggestion:**
Im not sure how crazy we need to go with testing but some edge cases I can think of:

- Self-closing tags with attributes (e.g., ``)
- Comments (`<!-- ... -->`)
- CDATA sections
- Malformed/mismatched tags
- Tags with complex attributes (quotes, special characters)

File: api/src/util/string_utils.py [hunk 10]
```diff
@@ -20,3 +35,88 @@ def is_valid_uuid(value: str) -> bool:
...
+        token = match.group(0)  # substring matched by regex
+        output_tokens.append(token)
+
+        if token.startswith("<"):  # Open Tag
+            tag_name = re.match(
+                r"</"  # Matches the begining of closing tag of an element
+                r"(\w+)",  # Matches one or more word characters
+                token,
+            )
+            if tag_name and token[1] == "/":  # Closing tag
```
Comment (joshtonava) [inline api/src/util/string_utils.py]: **question:** The closing tag matching logic checks `token[1] == "/"` after already matching with a regex that starts with r"</". This seems redundant. Is the double check intentional for safety, or can it be simplified?

File: api/src/util/string_utils.py [hunk 11]
```diff
@@ -20,3 +35,88 @@ def is_valid_uuid(value: str) -> bool:
         return True
     except ValueError:
         return False
+
+
+def contains_regex(value: str, regex: str) -> bool:
```
Comment (joshtonava) [inline api/src/util/string_utils.py]: **nit:** This is only used once, consider prefixing with an underscore for internal use or even inlining.

File: api/src/util/string_utils.py [hunk 12]
```diff
@@ -20,3 +35,88 @@ def is_valid_uuid(value: str) -> bool:
...
+        r"[^<]+"  # Match one or more characters (excluding <)
+    )
+
+    for match in tag_regex.finditer(html_str):
+        token = match.group(0)  # substring matched by regex
+        output_tokens.append(token)
+
+        if token.startswith("<"):  # Open Tag
+            tag_name = re.match(
+                r"</"  # Matches the begining of closing tag of an element
```
Comment (joshtonava) [inline api/src/util/string_utils.py]:
```suggestion
                r"</"  # Matches the beginning of closing tag of an element
```

File: api/src/util/string_utils.py [hunk 13]
```diff
@@ -20,3 +35,88 @@ def is_valid_uuid(value: str) -> bool:
...
+        if closing_match and closing_match.group(1) in BLOCK_TAGS:
+            output_tokens.pop()
+        else:
+            break
+    # Close any remaining open tags (excluding block-level)
+    closing_tags = [f"</{tag}>" for tag in reversed(tag_stack) if tag not in BLOCK_TAGS]
+    return "".join(output_tokens) + "".join(closing_tags)
+
+
+def truncate_html_safe(value: str, max_length: int) -> str:
```
Comment (joshtonava) [inline api/src/util/string_utils.py]: **suggestion:** Consider adding a doc string here.

File: api/src/util/string_utils.py [hunk 14]
```diff
@@ -20,3 +35,88 @@ def is_valid_uuid(value: str) -> bool:
...
+                # Add non self-closing tags
+                tag_name = re.match(
+                    r"<"  # Matches the opening angle bracket of an HTML tag,
+                    r"(\w+)",  # Matches one or more word characters
+                    token,
+                )
+                if tag_name and not token.endswith("/>"):
+                    tag_stack.append(tag_name.group(1))
+
+    # Remove block-level closing tags from the end
```
Comment (joshtonava) [inline api/src/util/string_utils.py]: suggestion: A comment explaining why block tags are left open would help future maintainers.

File: api/src/util/string_utils.py (hunk 13 above)
Comment (chouinar) [inline api/src/util/string_utils.py]: I think this logic is getting quite complex / there's a lot we'd have to account for. Have we looked into any libraries that might handle this for us?

Something like https://stackoverflow.com/a/9205760 ?

File: api/src/util/string_utils.py (hunk 7 above)
Comment (babebe) [reply api/src/util/string_utils.py]: Now, using bs4 to handle truncation

File: api/tests/src/task/notifications/test_opportunity_notification.py
Comment (babebe) [reply api/tests/src/task/notifications/test_opportunity_notification.py]: udpated

File: api/src/util/string_utils.py (hunk 8 above)
Comment (babebe) [reply api/src/util/string_utils.py]: Added new func

File: api/src/util/string_utils.py (hunk 9 above)
Comment (babebe) [reply api/src/util/string_utils.py]: Now, using bs4 to handle truncation

File: api/src/util/string_utils.py (hunk 10 above)
Comment (babebe) [reply api/src/util/string_utils.py]: Now, using bs4 to handle truncation

File: api/src/util/string_utils.py (hunk 11 above)
Comment (babebe) [reply api/src/util/string_utils.py]: Removed

File: api/src/util/string_utils.py (hunk 13 above)
Comment (babebe) [reply api/src/util/string_utils.py]: added doc string to new func

Comment (babebe) [reply api/src/util/string_utils.py]: Rewrote logic to now, use bs4 to handle truncation.

File: api/src/util/string_utils.py [hunk 15]
```diff
@@ -20,3 +23,83 @@ def is_valid_uuid(value: str) -> bool:
         return True
     except ValueError:
         return False
+
+
+def contains_regex(value: str, regex: str) -> bool:
+    return bool(re.search(regex, value))
+
+
+def _strip_partial_tag(html: str) -> str:
```
Comment (joshtonava) [inline api/src/util/string_utils.py]: **suggestion:** If we aren't using this we should remove it

File: api/src/util/string_utils.py (hunk 11 above)
Comment (joshtonava) [reply api/src/util/string_utils.py]: is this removed?

File: api/src/util/string_utils.py (hunk 15 above)
Comment (babebe) [reply api/src/util/string_utils.py]: Removed

**Comment by babebe** [comment]
> I think this looks good. Can you maybe put something in the PR description for why beautiful soup was chosen?

Updated Description
